src/effection/events/mailbox.ts

```diff
@@ -0,0 +1,46 @@
+import { fork, monitor, Operation } from 'effection';
+import { EventEmitter } from 'events';
+
+import { on, onEach, EventName } from '../events';
+import { compile } from './pattern';
+export { any } from './pattern';
+
+export class Mailbox {
+  private subscriptions = new EventEmitter();
+  private messages = new Set();
+
+  *send(message: unknown): Operation {
+    this.messages.add(message);
+    this.subscriptions.emit('message');
+  }
+
+  *receive(pattern: unknown = undefined): Operation {
+    let match = compile(pattern);
+    let { messages, subscriptions } = this;
+    while (true) {
+      for (let message of messages) {
+        if (match(message)) {
+          messages.delete(message);
+          return message;
+        }
+      }
+      yield on(subscriptions, 'message');
+    }
+  }
+
+  static *watch(
+    emitter: EventEmitter,
+    events: EventName | EventName[],
+    prepare: (event: { event: string; args: unknown[] }) => unknown = x => x
+  ): Operation {
+    let mailbox = new Mailbox();
+    let parent = yield ({ resume, context: { parent }}) => resume(parent.parent);
+
+    parent.spawn(monitor(function* () {
```

Comment:
Why is this using `monitor` rather than `fork`?

Reply:
If we spawn a fork in the parent, then it will block the parent from continuing, which I don't think we want since in general, watching is forever:

```ts
yield Mailbox.watch(connection, "message");

return true;
```

We want the operation to finish after the return